JBMC report FAILURE on uncaught exception

peterschrammel opened the pull request:
some more regression tests and other minor fixes

peterschrammel committed:
Recompile regression test class files

The java files were changed in previous commits,
but the class files were not recompiled, which
caused confusing mismatches in line numbers.

tautschnig commented on regression/cbmc-java/enum2/test.desc:
```diff
@@ -0,0 +1,10 @@
+--
+^warning: ignoring
+--
+ordinal() not loaded by --lazy-methods
```
Should there be an issue for this, or maybe is there one already in some project?

tautschnig commented on regression/cbmc-java/enum3/test.desc:
```diff
@@ -0,0 +1,10 @@
+--
+^warning: ignoring
+--
+ordinal() not loaded by --lazy-methods
```
Should there be an issue?

tautschnig commented on regression/jbmc-strings/append_obj_int_tostring/test.desc:
```diff
@@ -0,0 +1,7 @@
+^EXIT=10$
+^SIGNAL=0$
+^VERIFICATION FAILED$
+--
```
Any chance to note which change/PR/commit has made this work?

tautschnig commented on src/java_bytecode/java_entry_point.cpp:
```diff
@@ -308,6 +308,16 @@ void java_record_outputs(
 
   init_code.move_to_operands(output);
+
+  // check that there is no uncaught exception
```
Shouldn't this be done in `goto_checkt` (in future, the same might be of interest for other languages), and is enforcing without giving any command-line control to the user a good idea?

tautschnig reviewed (commented):
1. I'm not a Java expert, but I think this warrants more work as indicated in my comments.
2. As GitHub doesn't allow me to comment on the changes in  the last commit ("Recompile regression test class files"): _why_ are these being recompiled?

peterschrammel committed:
Assert that there is uncaught exception

That's the default behaviour because an escaping exception
makes the JVM abort. The user can override this behaviour
using the --disable-uncaught-exception-check option.

peterschrammel commented:
@tautschnig, I've refined the behaviour now:
* Asserting that there is uncaught exception remains the default behaviour because an escaping exception makes the JVM abort; we should replicate that behaviour and report that as a bug. However, the user can now override this behaviour using the --no-uncaught-exception-check option, which might be useful in combination with --function, when thrown exceptions could be considered as part of the 'normal' behaviour.
* Moreover, I've introduced --propagate-assertion-error which allows to propagate AssertionError as performed in the JVM rather than using a goto ASSERT statement. This flag is turned off by default following the assumption that usually nobody tries to catch AssertionError, and hence reporting a failing assertion at the location where AssertionError is thrown is typically the desired behaviour; if not the flag can be used.

peterschrammel commented:
TG bump is passing.

tautschnig commented on jbmc/regression/jbmc/enum2/test.desc:
```diff
@@ -0,0 +1,10 @@
+--
+^warning: ignoring
+--
+ordinal() not loaded by --lazy-methods
```
Maybe this is tracked elsewhere, otherwise might be worth opening an issue.

tautschnig commented on jbmc/regression/jbmc/enum3/test.desc:
```diff
@@ -0,0 +1,10 @@
+--
+^warning: ignoring
+--
+ordinal() not loaded by --lazy-methods
```
As for the other one: might warrant an issue.

tautschnig commented on jbmc/regression/jbmc-strings/append_obj_int_tostring/test.desc:
```diff
@@ -0,0 +1,7 @@
+^EXIT=10$
+^SIGNAL=0$
+^VERIFICATION FAILED$
+--
```
Since this is a _regression_ test: when was this an issue, could the commit message maybe say something to that effect?

peterschrammel replied:
Would require a bisect between now and January.

tautschnig reviewed (commented):
No serious objections from my side (I'm not a Java person anyway), just procedural comments. In addition to the below: "Recompile regression test class files" - why?

thk123 commented on jbmc/src/java_bytecode/java_bytecode_instrument.cpp:
```diff
@@ -592,6 +593,30 @@ void java_bytecode_instrument_symbol(
+/// Instruments the start function with an assertion that checks whether
+/// an exception has escaped the entry point
+/// \param symbol_table: global symbol table
+void java_bytecode_instrument_uncaught_exceptions(symbol_tablet &symbol_table)
```
Is mutating the entry point a good plan? Might be better for the entry point construction to have a collection of `checks` that it appends after the return. Can't spot a concrete problem (so feel free to ignore) but I think strongly couples this code with the entry point generation (if the entry point code changes to wrap the whole thing in an if statement for whatever reason, the assert will be in the wrong place).

peterschrammel replied:
In my opinion, the entire property instrumentation should be moved into a goto-pass.

thk123 replied:
Be that as it may - is that likely to happen in the near future, and if not would decoupling be useful in the meantime?

peterschrammel replied:
I've added some invariants to check that the entry point looks sane.

thk123 commented on jbmc/src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -778,14 +775,20 @@ bool java_bytecode_languaget::generate_support_functions(
       object_factory_parameters,
-      get_pointer_type_selector());
+      get_pointer_type_selector()))
+    return true;
```
If statement is multiple lines so wrap the body in braces

thk123 commented on jbmc/src/java_bytecode/java_bytecode_language.h:
```diff
@@ -26,20 +26,23 @@ Author: Daniel Kroening, kroening@kroening.com
   "(java-no-load-class):"
 
 #define JAVA_BYTECODE_LANGUAGE_OPTIONS_HELP /*NOLINT*/                                          \
+  " --no-uncaught-exception-check    ignore uncaught exceptions and errors\n" \
```
You might consider documenting this feature (and this flag to turn it off) more thoroughly in a markdown file somewhere (`jbmc/doc/cl-options.md`?)

peterschrammel replied:
This would be useful. I've created a ticket for it (TG-3893).

thk123 replied:
Sorry should have pointed this out in first review, but given that the error message is "no uncaught exception", this flag reads a bit like it turns on the "no uncaught exception check", suggest calling this flag: `--disable-uncaught-exception-check`?

thk123 commented on jbmc/src/java_bytecode/java_bytecode_language.h:
```diff
@@ -43,6 +44,7 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #define JAVA_BYTECODE_LANGUAGE_OPTIONS_HELP /*NOLINT*/                                          \
   " --no-uncaught-exception-check    ignore uncaught exceptions and errors\n" \
+  " --propagate-assertion-error      propagate java.lang.AssertionError\n" \
```
Also this as I don't know what this means (propagate to who?)

thk123 commented on jbmc/src/java_bytecode/remove_exceptions.cpp:
```diff
@@ -383,10 +389,10 @@ void remove_exceptionst::instrument_throw(
-  // and this may reduce the instrumentation considerably if the programmer
-  // used assertions)
-  if(assertion_error)
+  // we allow AssertionError not to be propgated since
```
Typo: propagated

thk123 reviewed (commented):
Probably check at least one of the tests you modified produces the same output with the ` --no-uncaught-exception-check `. 

Add test for propagate-assertion-error

peterschrammel reviewed (commented):
@thk123, I added tests for both options as suggested.

peterschrammel commented:
>  In addition to the below: "Recompile regression test class files" - why?

Added explanation to commit message.

peterschrammel commented:
@thk123, I have taken into account your comments. TG bump is passing.

thk123 commented on jbmc/regression/jbmc-strings/append_obj_int_tostring/test.desc:
```diff
@@ -0,0 +1,7 @@
+^EXIT=10$
+^SIGNAL=0$
+^VERIFICATION FAILED$
+--
```
Needs a comment explaining why this is a `symex-driven-lazy-loading-expected-failure`

smowton replied:
Yeah that sounds odd for this one?

thk123 commented on jbmc/src/java_bytecode/java_bytecode_language.h:
```diff
@@ -26,19 +26,24 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #define JAVA_BYTECODE_LANGUAGE_OPTIONS_HELP /*NOLINT*/                                          \
+  " --no-uncaught-exception-check    ignore uncaught exceptions and errors\n" \
+  " --propagate-assertion-error      propagate java.lang.AssertionError as exception\n"          /* NOLINT(*) */ \
```
Again should have asked in the first review - but why isn't this the default behaviour?

thk123 replied:
I also suggest being more explicit here in lieu of `cl-options.md`:
```
treat user provided assertions as exceptions of type java.lang.AssertionError rather than jbmc asserts to try and violate
```

If that is accurate.

peterschrammel replied:
> why isn't this the default behaviour?

see commit message

thk123 commented:
Not sure why GH split up my review - nothing worth a re-review  from either but you should definitely address: https://github.com/diffblue/cbmc/pull/2043#discussion_r194311668

smowton commented on jbmc/src/java_bytecode/java_bytecode_instrument.cpp:
```diff
@@ -592,6 +593,30 @@ void java_bytecode_instrument_symbol(
+    symbol_table.lookup_ref(JAVA_ENTRY_POINT_EXCEPTION_SYMBOL);
+
+  symbolt &function =
+    symbol_table.get_writeable_ref(goto_functionst::entry_point());
```
Huh? Why not just do this in `java_entry_point`, which already builds the machinery to catch the toplevel exception and record it as an `OUTPUT`.

peterschrammel replied:
See comment by @thk123 above for keeping the generation of instrumentations in the same place.

smowton replied:
hmm. This seems very brittle at least, as it will break if the structure of __CPROVER_start changes incidentally. How about `java_entry_point` introduces a hook, `on_uncaught_toplevel_exception()` or something, which instrument populates?

peterschrammel replied:
I've added some invariants to check that the entry point looks sane. Adding a separate seemed an overkill.

smowton commented on jbmc/src/java_bytecode/remove_exceptions.cpp:
```diff
@@ -472,7 +478,21 @@ void remove_exceptionst::instrument_exceptions(
+          instr_it->guard.is_false() &&
+          instr_it->source_location.get_bool("user-provided"))
+        {
+          instr_it->make_skip();
```
Instead of fixing this up afterwards, how about doing it right the first time? This gets handled here (https://github.com/diffblue/cbmc/blob/10a468514c5b793f935e19029b66437586bedfe0/src/goto-programs/builtin_functions.cpp#L740), but this suggests the frontend should be in charge of adding the `ASSERT(FALSE)` or not as appropriate

peterschrammel replied:
Yes, but goto-programs is not the right place for adding more java-specifics and needs to be refactored.

smowton replied:
Right, that's what I'm saying. Instead of doing something weird here and then later `goto_convertt` special-cases `java::java.lang.AssertionError.<init>:`, why not remove the special case and have the frontend / java-instrument pass do the right thing (either generate an `ASSERT FALSE` or don't, depending on whether the user wants `AssertionError` to be propagated)

smowton reviewed (commented):
Suggest trying to synthesise the right code the first time, rather than do ad-hoc fixups as we go.

smowton commented on jbmc/src/java_bytecode/java_bytecode_instrument.cpp:
```diff
@@ -592,6 +593,30 @@ void java_bytecode_instrument_symbol(
+
+  symbolt &function =
+    symbol_table.get_writeable_ref(goto_functionst::entry_point());
+  codet &init_code = to_code(function.value);
```
At least check this is a `code_blockt`

peterschrammel committed:
Refactor user-defined assertion translation for Java

Assertions in Java are "throw a;" statements where
a is of type java.lang.AssertionError (an exception,
or Throwable, to be precise). Sometimes we want to
translate it into an ASSERT instruction in the goto
program. Special-casing in order to  handle that
was scattered across multiple classes. In this commit
we special-case it only once in the Java frontend
and translate it into assert(false); assume(false);
which is then correctly handled by later stages of the
translation.

peterschrammel committed:
Remove wrong assumption from goto check

The argument of throw might be null even
if it is of type java.lang.AssertionError.

peterschrammel committed:
Enable throwing of AssertionError

Introduces --throw-assertion-error which allows
to propgate AssertionError as performed in the JVM
rather than using a goto ASSERT statement.

peterschrammel commented:
@smowton, I've implemented your suggestion to clean up the user-defined assertion generation for Java right away. This looks much cleaner now and removes some references to `java.util.AssertionError` from `analyses` and `goto-programs`.

smowton commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -1237,12 +1240,51 @@ codet java_bytecode_convert_methodt::convert_instructions(
+        code_function_callt assert_call;
+        assert_call.function() =
+          symbol_exprt(CPROVER_PREFIX "assert", assert_type);
+        assert_call.lhs().make_nil();
```
You can simply use `code_assertt` and `code_assumet` instead of going via functions

smowton commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -1237,12 +1240,51 @@ codet java_bytecode_convert_methodt::convert_instructions(
-      c=code_expressiont(throw_expr);
-      results[0]=op[0];
+        code_blockt ret_block;
+        ret_block.operands().push_back(std::move(assert_call));
```
Unsure which we prefer of `std::move` vs. `move_to_operands` these days, but check

smowton reviewed (commented):
The new way of dealing with `throw new AssertionError` is much better :D Couple of minor improvements, and I still really dislike hacking `__CPROVER__start` rather than simply asking `java_entry_point` to build it correctly the first time.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 68cde168ca52cbf51147b12efc755f46fb23db7e).

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_instrument.cpp:
```diff
@@ -592,6 +593,27 @@ void java_bytecode_instrument_symbol(
+  assert_no_exception.assertion() = equal_exprt(
+    exc_symbol.symbol_expr(),
+    null_pointer_exprt(to_pointer_type(exc_symbol.type)));
+  source_locationt assert_location = source_location;
```
Is there actual value in this temporary?

peterschrammel replied:
It's shorter than `assert_no_exception.add_source_location().set_comment("no uncaught exception");`

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -777,15 +774,21 @@ bool java_bytecode_languaget::generate_support_functions(
 
   // generate the test harness in __CPROVER__start and a call the entry point
-  return
+  if(
```
Why is this `return` replaced by an `if`? Unless I'm missing something this code is `if(x) return true; return false;` - which should be the same as `return x;`

peterschrammel replied:
That was a leftover from the previous variant where the instrumentation was added from here. Cleaned up.

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -2248,10 +2251,39 @@ void java_bytecode_convert_methodt::convert_athrow(
+    // ASSUME false:
+    code_assertt assert_code;
+    assert_code.assertion() = false_exprt();
+    source_locationt assert_location = location; // copy
```
What is this copy good for?

peterschrammel replied:
Because we use `location` as the basis for a modified source location.

thk123 replied:
Nit: either explain that in the comment, or delete the comment (saying copy afterwards just invites the question)

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -2248,10 +2251,39 @@ void java_bytecode_convert_methodt::convert_athrow(
+
+    code_assumet assume_code;
+    assume_code.assumption() = false_exprt();
+    source_locationt assume_location = location; // copy
```
As above: why copy?

peterschrammel replied:
same here

tautschnig commented on src/analyses/goto_check.cpp:
```diff
@@ -1627,21 +1627,16 @@ void goto_checkt::goto_check(
-          notequal_exprt not_eq_null(
-            pointer,
-            null_pointer_exprt(to_pointer_type(pointer.type())));
+        notequal_exprt not_eq_null(
```
`const`

tautschnig reviewed (approved):
Just nit picks. Disclaimer: I have not done a technical review of the Java parts, just code reading.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 0b9334de9cde71bd2b553c70a81263c8e3d2b7c3).

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: cd2ef4b0700cb3581e0fa7f4e9e19721124fe0ef).
